scalableminds/webknossos
# Add Terms of Service Acceptance Concept

New config keys `webKnossos.termsOfService.{enabled, version, content}`, new routes
`GET /termsOfService`, `POST termsOfService/accept`, `GET /termsOfService/acceptanceNeeded`

The idea is that after login, the frontend checks acceptanceNeeded and shows a popup asking for new TOS acceptance. If the user is orga owner, they cann call `accept` and that’s it, if they are not, they are told `please contact your orga owner`.

Note that if termsOfService is not explicitly enabled in the config, `acceptanceNeeded` should always return false.

Note that when migrating an existing instance, every orga should get one organization owner.

### TODO Frontend
 -  [X] adapt types (user json now contains isOrganizationOwner)
 -  [X] use routes, show modals, as described above

### URL of deployed dev instance (used for testing):
- https://tos.webknossos.xyz

### Steps to test:
- set `termsOfService.enabled` to true in application.conf
- ensure that you are logged in
- open some page on wk
- a modal should appear and prompt you to accept the ToS
- log in as a second user (which is not an admin)
- a modal should appear and ask you to contact the orga admin
- snoozing
  - the default deadline is currently in 2023, so the modals should be closable
  - reload the page --> the modals shouldn't re-appear
  - wait more than 20 seconds --> the modals should appear again
- with exceeded deadline:
  - change the deadline in application conf to a past date (e.g., 2021)
  - check that the modals cannot be closed without hitting accept
  - after acceptance, the modals shouldn't appear again
- bump the ToS version in the application.conf
  - the modals should appear again

### Issues:
- fixes #6611 

------
(Please delete unneeded items, merge only when none are left open)
